### PR TITLE
1.52.1 - Blog post explaining the fingerprint issue.

### PR DESCRIPTION
**Pull request by pnkfelix**

### CODE REVIEW

**Review by Mark-Simulacrum — DISMISSED** [review]
I will likely aim to merge this in roughly an hour, but it looks good to me.

**Comment by Mark-Simulacrum** [comment]
Moving publish date to Monday morning.

File: posts/2021-05-07-caught-red-handed.md
Comment (pnkfelix) [reply posts/2021-05-07-caught-red-handed.md:88]: omg thank you!

File: posts/2021-05-07-caught-red-handed.md
Comment (pnkfelix) [inline posts/2021-05-07-caught-red-handed.md:101]: Is this language too strong? I'm reflecting on the feedback from @carols10cents [here](https://github.com/rust-lang/rust/issues/84970#issuecomment-834771043).

Should we be using the softer language of "We recommend users of 1.52.0 disable incremental compilation *if* they are encountering unstable fingerprints repeatedly" ? We still don't have a good understanding of how often people are running into this, and if they run into it repeatedly even after their first `cargo clean`...

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:105]: I'm not sure if yinz have had time to do this analysis yet, but I'd love to see some reflection in this section on "how did this get missed in nightly/beta and how would we prevent similar mistakes in the future", or an indication that this analysis will be done and published after the fixes are out.

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:8]: I think it would be really helpful to have a Summary/TL;DR section at the top that says something along the lines of:

- There are bugs in incremental compilation in 1.52.0 stable. 
- Incremental compilation is off by default for release builds, so no production builds should be affected. 
- If you are encountering these bugs during development, you can disable incremental compilation by using the environment variable `CARGO_INCREMENTAL=0` or editing your `$HOME/.cargo/config.toml` or `$HOME/.cargo/config.toml` file to add `build.incremental = 'false'`.
- We're expecting to have a 1.52.1 release on [whatever date]
- We [do|do not] want more bug reports (and if you do want them, what's the most useful information to include)

**Review by carols10cents — CHANGES_REQUESTED** [review]
Haven't finished reading yet but wanted to submit comments so far

File: posts/2021-05-07-caught-red-handed.md
Comment (est31) [reply posts/2021-05-07-caught-red-handed.md:8]: > There are bugs in incremental compilation in 1.52.0 stable.

*and prior versions*

File: posts/2021-05-07-caught-red-handed.md
Comment (Mark-Simulacrum) [reply posts/2021-05-07-caught-red-handed.md:105]: I'll provide some background and maybe you can help indicate which of this is useful/helpful information - it may be published in a retrospective later, but I don't know yet if that'll happen.

On the day of the release I submitted https://github.com/rust-lang/rust/issues/84970 to resolve the question of what to do on 1.53 and 1.54 releases. The verification change landed in 1.52, but I had mistakenly assumed that it had landed in 1.53 and we had an additional 6 weeks to either resolve it or improve our messaging; so, ultimately, the fact that this ICE existed *was* known for 12 weeks now, and indeed a few people have been driving the fixing. I think it was not as well understood how common it was - to my knowledge, we had not received reports until after the stable release of it occurring on *every* build, for example. So it was a priority, but not too strongly so.

Another part of the reason for the lack of prioritization is that this is a longstanding bug - to my knowledge, perhaps even to the initial addition of incremental (at least the red/green system, added a few years back), I don't think I'm aware of work since then that would have caused this. So, while it is true that the ICEs are definitely annoying, they merely reveal underlying problems - and in some sense, are better than nothing. I think everyone is in agreement that they are not the best user experience, and that we should avoid them if at all possible, but I think no one has come up with a straightforward solution to do so. It's possible that an alternative to deploying them would be to disable incremental as a default-on option, and indeed that is under active consideration for next week's point release to mitigate the problem.

In hindsight, it's possible that when we discovered this bug we should have immediately moved to do so. However, disabling such a milestone feature that *does* contribute significant performance improvements to many builds would be a hard sell, especially in light of some builds not triggering the verification panics - and as such being able to compile smoothly, even with the added verification. The tradeoff here is not clear.

I will also share that the release team discussed this last Friday and is expecting to invest in better tooling for tracking bugs (regressions in particular, but more generally too) as they propagate to stable, so that we can have a better awareness and hopefully avoid mistakes of the "woops, I thought this would be in 1.53" variety. This should help with a number of other challenges we've encountered as well. However, I expect this to be a long-term project, not one we can quickly address.

File: posts/2021-05-07-caught-red-handed.md
Comment (Mark-Simulacrum) [reply posts/2021-05-07-caught-red-handed.md:8]: > There are bugs in incremental compilation in 1.52.0 stable.

Note that there are bugs in incremental compilation ... in essentially all known versions, in 1.52 they're just very visible. I think we should be very clear that downgrading to 1.51 is likely *not* the path users want to take.

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [reply posts/2021-05-07-caught-red-handed.md:101]: If it would be helpful for folks to provide you with more information about what they're seeing, this would be a great place to give more details on the most helpful information and ways to provide that. I think some folks would be happy to put up with some inconvenience if they can help to identify and fix the problems sooner.

Some folks will definitely just want to get on with their work though, and it sucks that the way to do that will make compile times slower, so I think it's important to acknowledge that. Something like:

```suggestion
To completely avoid running into this problem when using 1.52.0, we recommend disabling incremental compilation via one of the methods described above until we have had a chance to identify and fix the problems and issue a new release. We know this will make compile times slower and we know that's not an ideal workaround, but we are working to make this as temporary a situation as possible.
```

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:103]: And we don't know how far back these bugs are present.

File: posts/2021-05-07-caught-red-handed.md
Comment (est31) [reply posts/2021-05-07-caught-red-handed.md:105]: > "how did this get missed in nightly/beta and how would we prevent similar mistakes in the future"

There has been bug filing going on basically ever since the ICEs were turned on.

If you check the [list of issues](https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+unstable+fingerprints) linked by @Mark-Simulacrum 's issue, you'll see the [oldest](https://github.com/rust-lang/rust/issues/83115) [issues](https://github.com/rust-lang/rust/issues/83126) having been filed on March 14 to the nightly compiler, with the [PR](https://github.com/rust-lang/rust/pull/83007) that enabled the checking having been merged in March 13, so right after that PR merged. With increased exposure to users, the number of reports increased.

But the bugs existed previously too, just caused silent corruptions and weird bugs instead, which is arguably worse than an explicit error message. To the contrary, turning on this error is "making sure the errors won't be missed any more".

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:111]: Committing to a date for the point release here, if that's at all possible, would go a long way. Even if you know it won't fix any problems, or will only fix a small number of problems, saying "we'll have a release with whatever fixes we have ready out" for people to try, and at that time saying whether we think there's going to be another point release or not, would make people feel like Something is Being Done.

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:125]: I think a more specific place for people to go who want to help with this specific problem would be better than rust-lang.org/community-- would the T-compiler stream on zulip be appropriate? WG-incr-comp? Maybe have a transient stream just for this issue so that the more general streams can be about other topics?

**Review by carols10cents — CHANGES_REQUESTED** [review]
Ok, thoughts on the post done. I'm happy to read revisions later today or tomorrow if that would be helpful!

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [reply posts/2021-05-07-caught-red-handed.md:8]: Is "There are bugs in incremental compilation in all Rust stable versions that were made into compilation errors in 1.52.0 stable." more accurate? I just edited my post to that

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [reply posts/2021-05-07-caught-red-handed.md:105]: Ok, I think the start of a TL;DR on root causes sounds to me like:

- We discovered a silent bug(s) with incremental compilation that have probably existed since incremental compilation was added
- We made the bug into a loud compiler error rather than a silent bug
- While that change was in nightly and beta, we got some bug reports but didn't realize exactly how often this bug would be hit in normal development with stable (most people/projects probably only run night/beta in CI where incremental compilation is either irrelevant or disabled)
- We thought this change was only going into beta this cycle; longer term, we will be looking into tooling improvements to provide more clarity of what version a change is in and possibly how many bugs are being reported about a change as a change switches trains
- It also sounds like we should do some thinking on how to get more user testing of parts of Rust only used during development while they're on nightly and beta?

Please feel free to edit this if I've misunderstood or if you disagree with my conclusions or save this stuff for a retro, but as a Rust user reading this, I mostly want to know that the team is going to take a bigger view of this as well as fixing the immediate issue.

File: posts/2021-05-07-caught-red-handed.md
Comment (camelid) [inline posts/2021-05-07-caught-red-handed.md:10]: I think you're missing an "s" here:

```suggestion
There was an instance of this in the most recent release, 1.52.0, which added a new bit of internal-consistency checking, called "incremental compilation hash verification" (abbreviated `verify-ich`). This check is also called an "unstable fingerprint" check, because the diagnostic it currently prints looks [like this](https://github.com/rust-lang/rust/issues/84336):
```

**Comment by Mark-Simulacrum** [comment]
Just pushed an update in light of my expectation that we'll be releasing a point release on Monday - see https://github.com/rust-lang/rust/pull/85097 - but I expect to continue working on the language tomorrow, thank you everyone for the comments so far.

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:10]: From what I understand of previous comments yinz have made, the bug(s) weren't *introduced* in 1.52.0 so much as *surfaced* or *made into a compiler error*. I'd also mention incremental compilation in the first sentence. So perhaps something like:

> The Rust team has prepared a new release, 1.52.1, working around a bug in incremental compilation that was made into a compiler error in 1.52.0.

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:11]: This sentence needs another comma to end the appositive phrase:

> We recommend all Rust users, including those currently using stable versions prior to 1.52.0, upgrade to 1.52.1.

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:28]: I would say "disables incremental compilation by default" here.

File: posts/2021-05-07-caught-red-handed.md
Comment (Mark-Simulacrum) [reply posts/2021-05-07-caught-red-handed.md]: @carols10cents It might be helpful to clarify what the distinctions between these things are - the *compiler* itself has no UB, but other than that the miscompilations could include any of these. The known case, for example, caused the wrong pointer to be loaded from a vtable (due to one piece of code expecting one layout for it, and another a different layout, IIRC) and called, leading to memory unsafety (and eventually a crash).

I think we could say more about what we mean by miscompilation but I'd like to better understand what you think the distinctions between these cases are so we can better explain it.

**Comment by Mark-Simulacrum** [comment]
Hey @carols10cents I applied a bunch of feedback from your comments I think, though some of them are also a bit stale after some movement - I would really love a review if you have a chance to do so :)

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:31]: I would say "all Rust versions since 1.24.0, when incremental compilation was turned on by default", according to here: https://github.com/rust-lang/rust/blob/master/RELEASES.md#cargo-27, I can't find quickly where incremental compilation was first shipped in stable. Effectively, this is "all Rust versions" because people usually upgrade, but I think it helps to be more specific if we can.

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:35]:
```suggestion
environment if on a prior version: please see the [what you should do][part3]
```

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:70]: I don't think it hurts to stress repeatedly that this is a bug in *incremental compilation*, not *all* compilation:

```suggestion
This is the error caused by the internal consistency check, and as stated in the diagnostic, it yields an "Internal Compiler Error" (or ICE). In other words, it represents a bug in the internals of the Rust compiler itself. In *this* case, the ICE is revealing a bug in incremental compilation that predates the 1.52.0 release and could result in miscompilation if it had not been caught by `verify-ich`.
```

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:138]: 👍🏻 👍🏻 👍🏻 👍🏻

File: posts/2021-05-07-caught-red-handed.md
Comment (carols10cents) [inline posts/2021-05-07-caught-red-handed.md:57]: 👍🏻 👍🏻 👍🏻 👍🏻

**Review by carols10cents — CHANGES_REQUESTED** [review]
This is lots, lots better!!!

I think the only bits of my first round of feedback that I don't see incorporated are:

- At least a short statement that we will be investigating root causes of how this change got to stable before we expected it to and how to prevent similar situations in the future
- At least a short definition of what "miscompilation" means to give a sense of what kind of "wrong" the compiler may have been silently generating

Without that information, I would bet money that those two topics will be the ones that get the most comments/questions when this is posted.

File: posts/2021-05-07-caught-red-handed.md
Comment (Mark-Simulacrum) [reply posts/2021-05-07-caught-red-handed.md:31]: Rust 1.24 is correct - I'm going to add this as a footnote, so we're not distracting too much from the central message. It was all the way back in February 2018, so my guess is most users have migrated.

**Comment by Mark-Simulacrum** [comment]
Hopefully addressed all remaining points. Thanks!

File: posts/2021-05-07-caught-red-handed.md
Comment (est31) [inline posts/2021-05-07-caught-red-handed.md:163]:
```suggestion
But regardless of whether or not you file a bug, the problem can be worked around on your end by either:
```

File: posts/2021-05-07-caught-red-handed.md
Comment (est31) [inline posts/2021-05-07-caught-red-handed.md:168]:
```suggestion
 3. forcing incremental compilation to be disabled, by setting `CARGO_INCREMENTAL=0` in your environment or `build.incremental` to `false` in the `config.toml`.
```

upgrading, deleting, forcing :)

**Review by carols10cents — APPROVED** [review]
This looks great now (with the few small edits @est31 pointed out that I agree with). Thank you so much for working on this ❤️

**Review by steveklabnik — DISMISSED** [review]
Thanks for doing such a thorough writeup on this. Looks good to me.

File: posts/2021-05-07-caught-red-handed.md
Comment (wesleywiser) [inline posts/2021-05-07-caught-red-handed.md:176]: For these users who are willing to put up with some ICEs, would it make sense to ask them to install the 1.53 beta? We have fixed a number in beta although not all of them.

File: posts/2021-05-07-caught-red-handed.md
Comment (wesleywiser) [inline posts/2021-05-07-caught-red-handed.md:212]: I think this raises the question "will incremental be disabled on beta?"

It might be good to preemptively address that question.

**Review by wesleywiser — DISMISSED** [review]

**Comment by Mark-Simulacrum** [comment]
@wesleywiser I think I'm currently inclined to leave beta and nightly in the non-disabled incremental states, but it's definitely a good idea to call out their status. I think it's not unlikely 1.53 will need to also ship without incremental, though.

File: posts/2021-05-07-caught-red-handed.md
Comment (pnkfelix) [reply posts/2021-05-07-caught-red-handed.md]: (This change is to emphasize that the the subject of the sentence is *not* version 1.52.0, which was referenced in the previous sentence.)

File: posts/2021-05-07-caught-red-handed.md
Comment (pnkfelix) [reply posts/2021-05-07-caught-red-handed.md:147]: mm, good catch; i hadn't thought of the scenario of adjusting the settings for individual profiles in a local manner.

File: posts/2021-05-07-caught-red-handed.md
Comment (pnkfelix) [reply posts/2021-05-07-caught-red-handed.md:176]: (The text a few paragraphs down does suggest that option, of trying out the 1.53 beta since a number of bugs are fixed there.)

File: posts/2021-05-07-caught-red-handed.md
Comment (pnkfelix) [reply posts/2021-05-07-caught-red-handed.md:173]: (I think the text in the summary does a good job of explaining that aspect of things. I don't think we need to reiterate the point here.)

**Review by wesleywiser — APPROVED** [review]

File: posts/2021-05-07-caught-red-handed.md
Comment (pietroalbini) [inline posts/2021-05-07-caught-red-handed.md:5]: Should this be the release team as this is a release?

**Review by pietroalbini — CHANGES_REQUESTED** [review]
Looks good!

We need to rename the file to `posts/2021-05-10-Rust-1.52.1.md`, as the date is wrong and the blog post slug is not the one we usually use for releases.

**Review by pietroalbini — APPROVED** [review]
Ready to merge as soon as we release!

**Pull request merged**
